Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -92,53 +140,76 @@ public final class HTTPServerPipelineHandler: ChannelDuplexHandler, RemovableCha
         /// never suppress reads again.
         case sentCloseOutput
 
-        mutating func requestHeadReceived() {
+        /// The user has violated an invariant. We should refuse further IO now
+        case preconditionFailed
+
+        mutating func requestHeadReceived() -> ConnectionStateAction {
             switch self {
+            case .preconditionFailed:
+                return .forceCloseConnection
             case .idle:
                 self = .requestAndResponseEndPending
+                return .none
             case .requestAndResponseEndPending, .responseEndPending, .requestEndPending,
                     .sentCloseOutputRequestEndPending, .sentCloseOutput:
-                preconditionFailure("received request head in state \(self)")
+                let message = "received request head in state \(self)"
+                self = .preconditionFailed
+                return .warnPreconditionViolated(message: message)
```

Comment:
We expect the user to close the channel, and we will end up closing it if they do not (because they will trigger another such error). However, it is better to do this in a way that hits the user's error handling code, as that will help surface it to them. They ought to have an error handler in the pipeline, and that error handler will log. If they do not, if they have an async channel, then the inbound stream will throw. All of these are ideal behaviours.

We should only close _for_ the user if it is clear that they are not recovering from the situation themselves.